### PR TITLE
Develop self play first prototype

### PR DESCRIPTION
**Pull request by Danielhp95**
First working prototype of the self-play system. 

Features:
Ghost brains follow a main brain
Keeps track of historical policies from main brain
Ghost brains sample from policy pool every now and again (hyper parameter)
Latest policy is always available to sample as a ghost policy

Bad features
It still relies on brain names to identify self-play brains.

### CODE REVIEW

File: python/unitytrainers/self_play_trainer_controller.py [hunk 1]
```diff
@@ -67,41 +70,62 @@ def create_variable_mapping(self, from_scope, to_scope):
...
+            if self.should_change_ghost_model():
+                self.resample_all_ghosts(sess)
+        return super(SelfPlayTrainerController, self).handle_episode_termination(curr_info, sess)
 
     def resample_all_ghosts(self, sess):
+        """
+        Samples a historical policy for all ghost trainers
+        :param sess: Tensorflow session.
+        """
+        self.store_main_latest_policy(sess, self.main_trainer_saver, policy_path=self.latest_policy_checkpoint_path)
```
Comment (Pete-York) [inline python/unitytrainers/self_play_trainer_controller.py:99]: Do we need `policy_path=`? Isn't this implicit from the argument orer?

**Review by Pete-York — COMMENTED** [review]
Apart from the one comment I made it looks good!

File: python/unitytrainers/self_play_trainer_controller.py (hunk 1 above)
Comment (Danielhp95) [reply python/unitytrainers/self_play_trainer_controller.py:99]: I wrote it purely for readability purposes, the interpreter can infer what it is by argument order. I would rather keep it, if it's all the same

Comment (Pete-York) [reply python/unitytrainers/self_play_trainer_controller.py:99]: I don't feel strongly about it but I'm not sure what's particular about this argument that makes you want to use it here and not in other locations.

Comment (Danielhp95) [reply python/unitytrainers/self_play_trainer_controller.py:99]: Would you prefer if we also explicitly stated the parameter names? I actually feel that for this highly technical academic code it is useful to be explicit about the operations being made.

**Pull request merged**